Source/Control/adsr.cpp

```diff
@@ -3,79 +3,77 @@
 
-float Adsr::Tau2Pole(float tau)
+
+void Adsr::Init(float sample_rate, int blockSize)
 {
-    return expf(-1.0f / (tau * sample_rate_));
+    sample_rate_ = sample_rate / blockSize;
+    SetTime(ADSR_SEG_ATTACK,  0.1f);
+    SetTime(ADSR_SEG_DECAY,   0.1f);
+    SetTime(ADSR_SEG_RELEASE, 0.1f);
+    sus_level_  = 0.7f;
+    x_          = 0.0f;
 }
 
-float Adsr::AdsrFilter()
+void Adsr::Retrigger(bool hard)
 {
-    y_ = b_ * x_ + a_ * y_;
-    return y_;
+    mode_ = ADSR_SEG_ATTACK;
+    if(hard)
+        x_ = 0.f;
 }
 
-
-void Adsr::Init(float sample_rate)
+void
+Adsr::SetTime(int seg, float time)
 {
-    seg_time_[ADSR_SEG_ATTACK]  = 0.1f;
-    seg_time_[ADSR_SEG_DECAY]   = 0.1f;
-    sus_                        = 0.7f;
-    seg_time_[ADSR_SEG_RELEASE] = 0.1f;
-    //timer_ = 0;
-    a_           = 0.0f;
-    b_           = 0.0f;
-    x_           = 0.0f;
-    y_           = 0.0f;
-    prev_        = 0.0f;
-    atk_time_    = seg_time_[ADSR_SEG_ATTACK] * sample_rate;
-    sample_rate_ = sample_rate;
+    if(seg_time_[seg] != time)
+    {
+        seg_time_[seg] = time;
+        if(time > 0.f)
+        {
+            const float attackTarget = logf(0.5f);
+            const float decayTarget  = logf(1./M_E);
+            float target = (seg == ADSR_SEG_ATTACK)? attackTarget : decayTarget;
+            seg_D0_  [seg] = 1.f - expf(target / (time * sample_rate_));
+        }
+        else
+            seg_D0_  [seg] = 1.f; // instant change
+    }
 }
 
+
 float Adsr::Process(bool gate)
 {
-    float pole, out;
-    out = 0.0f;
+    float out = 0.0f;
 
     if(gate && mode_ != ADSR_SEG_DECAY)
-    {
         mode_ = ADSR_SEG_ATTACK;
-        //timer_ = 0;
-        pole      = Tau2Pole(seg_time_[ADSR_SEG_ATTACK] * 0.6f);
-        atk_time_ = seg_time_[ADSR_SEG_ATTACK] * sample_rate_;
-        a_        = pole;
-        b_        = 1.0f - pole;
-    }
     else if(!gate && mode_ != ADSR_SEG_IDLE)
-    {
         mode_ = ADSR_SEG_RELEASE;
-        pole  = Tau2Pole(seg_time_[ADSR_SEG_RELEASE]);
-        a_    = pole;
-        b_    = 1.0f - pole;
-    }
-
-    x_    = (int)gate;
-    prev_ = (int)gate;
 
+    float D0(seg_D0_[mode_]);
+    float target = mode_ == ADSR_SEG_DECAY? sus_level_ : - 0.1f;
     switch(mode_)
     {
         case ADSR_SEG_IDLE: out = 0.0f; break;
         case ADSR_SEG_ATTACK:
-            out = AdsrFilter();
-
-            if(out > .99f)
+            x_ += D0 * (2.f - x_);
+            out = x_;
+            if(out > 1.f)
             {
+                x_  =
+                out = 1.f;
                 mode_ = ADSR_SEG_DECAY;
-                pole  = Tau2Pole(seg_time_[ADSR_SEG_DECAY]);
-                a_    = pole;
-                b_    = 1.0f - pole;
             }
             break;
         case ADSR_SEG_DECAY:
         case ADSR_SEG_RELEASE:
-            x_ *= sus_;
-            out = AdsrFilter();
-            if(out <= 0.01f)
+            x_ += D0 * (target - x_);
+            out = x_;
+            if(out < 0.0f)
```

Comment:
In my tests, `x_` never reaches 0. So the ADSR ends up in an infinite `DECAY` state.


Probably it would make sense to set it to something like that:

```suggestion
            if(out < 0.01f)
```

Reply:
No, then we would get a glitch again, where the output jumps from 0.01f to 0.0f suddenly. Its better set the target to something smaller than zero for the `RELEASE` stage. @sdiedrichsen is doing that already, so I'm not sure why you see it never reaching 0.0f?

Reply:
I'm not sure tbh. I basically copy-pasted two files into my project, and ran it as is. Adsr stuck in decay phase. In my debugging attempts I noticed that _x gets indefinitely small, but never reaches 0. 

I'll try from a clean state tomorrow, but I doubt it's something in my code tbh.

Reply:
Here's a gist: https://gist.github.com/brbrr/9155763905f54b3ceb66c838c3db1e4b

Reply:
barber, thanks for taking your time and testing the code. 
The observation is true and intended. As long as the gate is TRUE and sustain level is 0.f, x_ reaches 0.f eventually, but this takes a very loooonnnng time.
And as TheSlowGroth mentioned, we'll have back our old friend, the glitch. 
Actually, an analog ADSR envelope has only three time related phases or states, Attack, Decay and Release. Sustain is not a phase, just a level, because you can't put a good number on it, as you saw in your experiment. If you need a method indicating the end of the decay phase, we might add a specific comparison for just that to give the user control options.

Reply:
Gotcha. Thanks for the explanation. 

After a bit more debugging, I figured what's my problem is. And it is indeed problem with my code :D 

In my project, I have a Metronome clock, that produces a SINE plucks every tick (it uses Osc running on BPM frequency for actual clock). There, I use `GetCurrentSegment() == 0` conditional to wait for a "previous" pluck to complete. Probably my design decision isn't great, but I can't really figure out how to properly wait for Env to fade out.

Here's a chunk of logic I mentioned above:

```
  if (metroEnv.GetCurrentSegment() == 0)
  {
    if (mod == 0 && beatGate == false)
		beatGate = true;
    else
		beatGate = false;
  }

  float metroAmp = metroEnv.Process(beatGate);
  metro.SetAmp(metroAmp);
  return metro.Process();
  ```

Reply:
Anyway, good catch. The API behavior has been modified, which is not good. But the solution might be easy:
In SetSustainLevel, we could simply introduce a special case:
if (sus_level == 0.f)
  sus_level = -0.01f; // a slight negative value to force the envelope into idle state.
Does that sound like a concept?